File: Go/TAD.go
package main

import "fmt"

type tKey int

type tNodeA struct{
	key tKey
	left,right ABB
}

type ABB *tNodeA

func error(s string){
	fmt.Printf(s)
}

func createNodeA(a ABB){
	a = new(tNodeA)
	if a == nil {
		error("  *** abb.createNodeA : No memory\n")
	}
}

func emptyTree(a ABB) {
	a = nil
}

func insert_r(a ABB, key tKey){
	if a == nil{
		createNodeA(a)
		a.key = key
		a.left = nil
		a.right = nil
	}else if key < a.key{
		insert_r(a.left,key)
	}else {
		insert_r(a.right,key)
	}
}

func insertKey(a ABB , key tKey){
	insert_r(a,key)
}

func leftSon(a ABB) ABB{
	return a.left
}

func rightSon(a ABB) ABB{
	return a.right
}

func key(a ABB) tKey{
	return a.key
}

func isEmptyTree(a ABB) bool{
	return a == nil
}

func preorderWithBracket(a ABB) {
	fmt.Printf("(")
	if ! isEmptyTree(a){
		if ! isEmptyTree(leftSon(a)) || isEmptyTree(rightSon(a)){
			fmt.Printf(" ",key(a)," ")
			preorderWithBracket(leftSon(a))
			fmt.Printf(" ")
			preorderWithBracket(rightSon(a))
		} else {
			fmt.Printf("  %d%s",key(a)," ")
		}	
	}
	fmt.Printf(")")
}

func main() {
	var abb ABB
	emptyTree(abb)
	insertKey(abb,4)
	preorderWithBracket(abb)
}

File: P1/Rust/TAD.rs
type Link = Option<Box<Node>>;

struct Node{
    root: i32,
    left: Link,
    right: Link,
}

fn createNodeA(key: i32)-> Link{
    Some(Box::new(Node{root:key, left:None, right:None}))
}

fn emptyTree() -> Link {
    return None;
}

fn insert_r(mut abb: Link, key:i32)->Link{
    match abb {
        Some(ref abb) =>{ if key < abb.root {
            insert_r(abb.left,key)
        } else if key > abb.root{
            insert_r(abb.right,key)
        }else{
            
        }},
        None => createNodeA(key),
    }
}

fn key(abb:Link)-> i32{
    match abb {
        Some(ref abb) => abb.root,
        None => panic!("ERROR: EmptyTree"),
    }
}
//fn emptyTree()->

fn main(){
    let mut abb = emptyTree();
    abb = insert_r(abb,4);
    println!("{}",key(abb));
}